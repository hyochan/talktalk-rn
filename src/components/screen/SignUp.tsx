import React, { Component, useState, useEffect } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  ScrollView,
} from 'react-native';
import { NavigationScreenProp, NavigationStateRoute } from 'react-navigation';

import styled from 'styled-components/native';

import TextInput from '../shared/TextInput';
import Button from '../shared/Button';
import StatusBar from '../shared/StatusBar';

import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

const StyledContainer = styled.View`
  flex:1;
  background-color: white;
  flex-direction: column;
  align-items: center;
`;

const StyledScrollView = styled.ScrollView`
  align-self: stretch;
`;

const StyledWrapper = styled.View`
  margin-top: 40;
  width: 78%;

  flex-direction: column;
  align-items: center;
`;

const StyledButtonWrapper = styled.View`
  margin-top: 24;
  width: 100%;
  align-items: flex-end;
`;

interface IProps {
  navigation: any;
}

interface IState {
  isRegistering: boolean;
  email: string;
  pw: string;
  displayName: string;
  statusMsg: string;
}

function Screen(props: IProps, state: IState) {
  const [isRegistering, setIsRegistering] = useState(false);
  const [email, setEmail] = useState('');
  const [pw, setPw] = useState('');
  const [displayName, setDisplayName] = useState('');
  const [statusMsg, setStatusMsg] = useState('');

  useEffect(() => {
    if (isRegistering) {
      setIsRegistering(false);
    }
  }, [isRegistering]);

  const onRegister = () => {
    setIsRegistering(true);
  };

  const onTextChanged = (type: string, text: string) => {
    switch (type) {
      case 'EMAIL':
        setEmail(text);
        break;
      case 'PW':
        setPw(text);
        break;
      case 'NAME':
        setDisplayName(text);
        break;
      case 'STATUS_MSG':
        setStatusMsg(text);
        break;
    }
  };

  return (
    <StyledContainer>
      <StatusBar isDarkContent={true}/>
      <StyledScrollView
        contentContainerStyle={{
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <StyledWrapper>
          <TextInput
            testID='email_input'
            style={{ marginTop: 8 }}
            txtLabel={getString('EMAIL')}
            txtHint={ getString('EMAIL') }
            txt={ email }
            onTextChanged={ (text) => onTextChanged('EMAIL', text)}
          />
          <TextInput
            testID='pw_input'
            style={{ marginTop: 24 }}
            txtLabel={ getString('PASSWORD') }
            txtHint={ getString('PASSWORD') }
            txt={ pw }
            onTextChanged={ (text) => onTextChanged('PW', text)}
            isPassword={ true }
          />
          <TextInput
            testID='name_input'
            style={{ marginTop: 24 }}
            txtLabel={getString('NAME')}
            txtHint={ getString('NAME') }
            txt={ displayName }
            onTextChanged={ (text) => onTextChanged('NAME', text)}
          />
          <TextInput
            testID='status_input'
            style={{ marginTop: 24 }}
            txtLabel={getString('STATUS_MSG')}
            txtHint={ getString('STATUS_MSG') }
            txt={ statusMsg }
            onTextChanged={ (text) => onTextChanged('STATUS_MSG', text)}
          />
          <StyledButtonWrapper>
            <Button
              testID='register'
              isLoading={isRegistering}
              onPress={() => onRegister()}
            >{getString('REGISTER')}</Button>
          </StyledButtonWrapper>
        </StyledWrapper>
      </StyledScrollView>
    </StyledContainer>
  );
}

export default Screen;
