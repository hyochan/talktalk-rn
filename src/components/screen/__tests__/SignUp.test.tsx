import 'react-native';
import * as React from 'react';
import SignUp from '../SignUp';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

const props = {
  navigation: {
    navigate: jest.fn(),
  },
};
const component: React.ReactElement = <SignUp {...props}/>;

describe('[SignUp] rendering test', () => {
  it('renders as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('[SignUp] interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let testingLib: any;

  beforeAll(() => {
    rendered = renderer.create(component);
    testingLib = render(component);
  });

  it('should changeText when email changed', () => {
    const emailInput = testingLib.getByTestId('email_input');
    emailInput.props.onTextChanged('email test');
    expect(emailInput.props.txt).toEqual('email test');
  });

  it('should changeText when password changed', () => {
    const passwordInput = testingLib.getByTestId('pw_input');
    passwordInput.props.onTextChanged('pw test');
    expect(passwordInput.props.txt).toEqual('pw test');
  });

  it('should changeText when name changed', () => {
    const nameInput = testingLib.getByTestId('name_input');
    nameInput.props.onTextChanged('name test');
    expect(nameInput.props.txt).toEqual('name test');
  });

  it('should changeText when status changed', () => {
    const statusInput = testingLib.getByTestId('status_input');
    statusInput.props.onTextChanged('status test');
    expect(statusInput.props.txt).toEqual('status test');
  });

  it('should simulate when [onRegister] clicked', () => {
    const registerBtn = testingLib.getByTestId('register');
    fireEvent(registerBtn, 'press');
    expect(registerBtn.props.isLoading).toEqual(false);
    registerBtn.props.onPress();
    expect(registerBtn.props.isLoading).toEqual(true);
  });
});
