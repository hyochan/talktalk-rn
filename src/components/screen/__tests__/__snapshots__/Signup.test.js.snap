// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders as expected 1`] = `
<Styled(View)>
  <Shared
    isDarkContent={true}
  />
  <Styled(ScrollView)
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Styled(View)>
      <Shared
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 8,
          }
        }
        txt=""
        txtHint="Email"
        txtLabel="Email"
      />
      <Shared
        isPassword={true}
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 24,
          }
        }
        txt=""
        txtHint="Password"
        txtLabel="Password"
      />
      <Shared
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 24,
          }
        }
        txt=""
        txtHint="name"
        txtLabel="name"
      />
      <Shared
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 24,
          }
        }
        txt=""
        txtHint="Status message"
        txtLabel="Status message"
      />
      <Styled(View)>
        <Button
          activeOpacity={0.5}
          height={60}
          id="register"
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          isWhite={false}
          onPress={[Function]}
          width={136}
        >
          Register
        </Button>
      </Styled(View)>
    </Styled(View)>
  </Styled(ScrollView)>
</Styled(View)>
`;
